Add Style Guide and API Design Guide

j9liu opened the pull request:
Fixes #1540, #1541

Probably not exhaustive, but hopefully a good start. There's the possibility of merging these into a single Coding Guide (similar to CesiumJS), but I'm not sure what the call is. I think there's enough separation for now, and we can always continue to add to these docs in the future.

kring commented on Documentation/api-design-guide.md:
````diff
@@ -0,0 +1,461 @@
+}
+```
+
+In `Serialize`, you can retrieve the version associated with the `UObject` and implement compatibility measures. Here, the deprecated value of `UseWebMercatorProjection` (which has been redirected by `[CoreRedirects]`) is used to set the new property as needed.
````
At the time `Serialize` is called, properties that refer to other objects will not reliably be set, and attempting to set one risks it being overwritten later in the load process. So if any of that is necessary, the solution is to also override `PostLoad` and use `GetLinkerCustomVersion` to determine if upgrading is necessary. We have a few examples of this.

kring reviewed (commented):
This is great @j9liu, thanks for writing it up! Sorry for the large number of small comments here.

j9liu commented on Documentation/api-design-guide.md:
```diff
@@ -0,0 +1,461 @@
+- [Blueprints](#blueprints)
+- [Deprecation and Backwards Compatibility](#deprecation-and-backwards-compatibility)
+
+## UObject Design
```
Whoof yeah, that's a typo. Thanks for the find!

j9liu commented on Documentation/api-design-guide.md:
```diff
@@ -0,0 +1,461 @@
+3. Add `BlueprintGetter` and `BlueprintSetter` functions.
+4. Override the `PostEditChangeProperty` method on the class. This allows it to be notified of changes to the property in the Editor.
+
+There may be a rare case where no action is necessary after setting a certain property from C++. In this case—and as long as this behavior is unlikely to change—the property can be declared in the `public:` section. Such a property is not likely to need `PostEditChangeProperty` or `BlueprintSetter` either.
```
Makes sense. This is written `developing-uobjects.md` and I wasn't sure if we still did this, but I agree with your judgement

azrogers commented on Documentation/style-guide.md:
```diff
@@ -0,0 +1,197 @@
+
+Before diving into this guide, we recommend the following:
+
+1. Browse Cesium Native's [C++ style guide](https://github.com/CesiumGS/cesium-native/blob/main/doc/style-guide.md) to familiarize yourself with fundamental principles and expectations for C++ development.
```
Yes! Might need to grab the latest version of the tag file (from https://cesium.com/learn/cesium-native/ref-doc/cesium-native.tag), but you should be able to replace this with `\ref style-guide "C++ style guide"`).

j9liu replied:
Would it be okay to just link to the guide as published online? (https://cesium.com/learn/cesium-native/ref-doc/style-guide.html)

Personally I still look at Github for our documentation so I appreciate when things can also work well on Github. But I'm willing to concede and use `\ref`

j9liu commented on Documentation/api-design-guide.md:
````diff
@@ -0,0 +1,461 @@
+  double MaximumScreenSpaceError = 16.0;
+  ```
+
+- The `EditAnywhere` specifier allows the property to show up in the Details panel of the Unreal Editor. The user can thus modify the value directly.
````
Will reword this so it's clearer. 👍

j9liu commented on Documentation/api-design-guide.md:
```diff
@@ -0,0 +1,461 @@
+
+- The `EditAnywhere` specifier allows the property to show up in the Details panel of the Unreal Editor. The user can thus modify the value directly.
+- The `BlueprintGetter` and `BlueprintSetter` define specific functions for getting and setting the value in Blueprints (see [UFunctions](#ufunctions)).
+- The `Category` indicates how the property should be organized in the Details panel. It appears under the "Level of Detail" category, which itself is nested in the "Cesium" category.
```
I'll mention it in "Quick Reference"!

j9liu commented on Documentation/api-design-guide.md:
````diff
@@ -0,0 +1,461 @@
+```c++
+// Forward declare the class with the UE_DEPRECATED macro.
+struct UE_DEPRECATED(
+    5.0,
````
That makes more sense. I copied this directly from code, so it looks like I messed up then 😅

j9liu commented on Documentation/style-guide.md:
```diff
@@ -0,0 +1,197 @@
+
+Generally, the above pointers are used whenever you instantiate something that inherits from an Unreal type (e.g., `UObject`, `USceneComponent`). Of course, raw pointers may be used where ownership is not required.
+
+Unreal also provides `TUniquePtr` and `TSharedPtr` definitions, equivalent to their `std` counterparts. For consistency, Cesium for Unreal prefers using `TUniquePtr` and `TSharedPtr`. This is especially important when the type `T` is a `UObject`.
```
Whoops, not sure where I got this idea from.

j9liu commented on Documentation/style-guide.md:
````diff
@@ -0,0 +1,197 @@
+From a glance, it is not clear what the intended frame-of-reference is. (Maybe you assumed it was in Unreal space due to "`Location`", but what is `Transform` for?) The function body might give us a hint, but the declaration could use more clarity. Adding documentation will definitely help here, and in some cases that is enough to avoid ambiguity.
+
+```c++
+/**
````
Hm, is there something about it that is interfering with Doxygen? (Maybe the square brackets?)

j9liu commented on Documentation/api-design-guide.md:
```diff
@@ -0,0 +1,461 @@
+| ---- | ----- | --------------- |
+| `BlueprintCallable` | Function can be executed in a Blueprint or Level Blueprint graph. | For functions that depend on a sequence of node execution. For example, `UCesiumEllipsoid::Create` requires that the ellipsoid be created before something is done on it.<br><br>`BlueprintCallable` results in an "execution pin" to indicate what happens after the Blueprint moves on. For `const` functions, no ...
+| `BlueprintPure` | Function can be executed in a Blueprint or Level Blueprint graph without an execution pin. | For functions whose execution does not require explicit sequencing logic, and for which it will not affect the owning object. For example, `UCesiumEllipsoid::ScaleToGeodeticSurface` will implicitly execute before its value is used, and it does not change the `UCesiumEllipsoid` itself,  ...
+| `BlueprintGetter` | Function can be used as a `BlueprintGetter` for a `UPROPERTY`. | Required to work with `BlueprintGetter` in `UPROPERTY`. Unreal complains if this is not present. |
```
Ah, this was the result from a flaky test. I removed `BlueprintGetter` from a `UFUNCTION` just to see what would happen, and it wasn't happy because there was no `Blueprint___` specifier on it. But replacing it with `BlueprintCallable` compiles fine as well. I'll fix this.

j9liu commented:
No apologies @kring, thank you for the thorough review! I _think_ I addressed all of your comments, except for the image resize that's breaking Doxygen. I'm happy just using Markdown and living with the bigger image.

I turned some of your comments into short new sections; let me know if I got everything right!

kring commented:
Thanks @j9liu! I've disabled `STRIP_CODE_COMMENTS` because the doxygen comments in the code blocks still weren't showing up. This doesn't seem to break anything in an obvious way, but we'll see how we go!
